cli/listener/src/main.rs

```diff
@@ -570,11 +570,41 @@ fn get_success_response(interval_prices: IntervalPrices) -> ResponseBox {
     .boxed()
 }
 
+/// Get a header with `Expires` key.
+/// If the end date for the query is far in the past we set it to expire in 1
+/// year. However, if the end date is in the same epoch, we set it to expire in
+/// the approximate time the next epoch begins.
+fn get_expires_header(
+    now: chrono::DateTime<chrono::Utc>,
+    current_slot: Slot,
+    interval_prices: &IntervalPrices,
+) -> Header {
+    let epoch_sched = EpochSchedule::without_warmup();
+    let (epoch_for_current_slot, _) = epoch_sched.get_epoch_and_slot_index(current_slot);
+
+    // Set the validity of the query until the next epoch
+    let expires_date = if epoch_for_current_slot == interval_prices.end_epoch {
+        let first_slot_next_epoch = epoch_sched.get_first_slot_in_epoch(epoch_for_current_slot + 1);
+        // Does not underflow because `first_slot_next_epoch <= current_slot`.
+        let slots_diff = first_slot_next_epoch - current_slot;
+        // Estimate for how many milliseconds the response is valid.
+        let ms_diff = slots_diff * SLOT_MS;
```

Comment:
We both know that `SLOT_MS` is a lie :joy: It’s defined to be 400ms, but in practice it’s more like 550–600ms. In this case it’s not an issue, it just causes us to underestimate how long the data will be valid for. But if you want to be more accurate, we could also base it on the observed block time, as

```rust
if interval_prices.duration_epochs() > 0 {
    let epoch_duration = interval_prices.duration_wall_time() / interval_prices.duration_epochs();
    let slot_duration = epoch_duration / epoch_sched.get_slots_in_epoch(interval_prices.end_epoch);
    now + slot_duration * slots_diff
}
```

We can even do `now + slot_duration * (slots_diff * 8 / 10)` to underestimate the duration again, so we don’t overshoot and cache for a long time when the epoch suddenly ends sooner than expected.

Also, we should have a minimum time as well. Even if we are very close to the epoch boundary, we don’t need to let the value expire instantly, we can cache it for 60 seconds or so. The APY is not *that* important that you need to know the new value within seconds after the new epoch starts.